### PR TITLE
Made a common interface for MvPlugin

### PR DESCRIPTION
**Pull request by NahuLD**
TODO:
- [x] Move messaging channel interfaces to common
- [x] Make utility for databases

### CODE REVIEW

**Comment by NahuLD** [comment]
I don't know if I'm too happy with that. Didn't want to change too many things.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 1]
```diff
@@ -59,7 +65,8 @@ public final void onLoad() {
    * @return Whether the load was successful.
    * @throws Exception Any error encountered upon loading.
    */
-  protected boolean load()
+  @Override
+  public boolean load()
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Use a default method in the parent interface instead.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 2]
```diff
@@ -97,7 +104,8 @@ public final void onEnable() {
    * @return Whether the enable was successful.
    * @throws Exception Any error encountered upon enabling.
    */
-  protected boolean enable()
+  @Override
+  public boolean enable()
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Use a default method in the parent interface instead.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 3]
```diff
@@ -129,67 +137,134 @@ public final void onDisable() {
    *
    * @throws Exception Any error encountered upon disabling.
    */
-  protected void disable()
+  @Override
+  public void disable()
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Use a default method in the parent interface instead.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 4]
```diff
@@ -129,67 +137,134 @@ public final void onDisable() {
...
+  @Override
   @NotNull
   public MinevictusUtilsBungee getMvUtil() {
     return MinevictusUtilsBungee.getInstance();
   }
 
+  @Override
   @NotNull
-  private BungeeCommandManager constructAcf() {
+  public BungeeCommandManager constructAcf() {
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: This is an implementation detail and should not be public by a parent interface.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 5]
```diff
@@ -129,67 +137,134 @@ public final void onDisable() {
...
   public final void registerCommands(@NotNull BaseCommand... commands) {
     var acf = getAcf();
     for (var command : commands) {
       acf.registerCommand(command);
     }
   }
 
+  @Deprecated
+  @Override
+  public final void registerCommands(boolean force, @NotNull BaseCommand... commands) {
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Forcing is a platform detail and should therefore not be common as it is not common in ACF.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 6]
```diff
@@ -293,14 +369,10 @@ public void setupCooldowns(@NotNull Database database, @NotNull AcfCooldowns[] c
    * @return The channel given.
    * @since 3.8.0
    */
+  @Override
   @NotNull
-  public <P, C extends PacketChannel<P>> C packetChannel(@NotNull C channel) {
+  public <P, C extends IPacketChannel<P, ?>> C packetChannel(@NotNull C channel) {
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Require a `PacketChannel<P>` here instead.

File: bungee/src/main/java/us/minevict/mvutil/bungee/utils/DatabaseUtils.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
+package us.minevict.mvutil.bungee.utils;
+
+import co.aikar.idb.*;
+import net.md_5.bungee.api.plugin.Plugin;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+
+/**
+ * Utilities ported over from Aikar's Bukkit IDB to manage DB creation for Bungee on a simpler manner.
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/utils/DatabaseUtils.java]: `in a simpler manner`.

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+public interface IMvPlugin<CM extends CommandManager<?, ?, ?, ?, ?, ?>, L> {
+    /**
+     * Called upon loading the plugin.
+     * <p>
+     * Returns true if the plugin successfully loads, false otherwise.
+     *
+     * @return Whether the load was successful.
+     * @throws Exception Any error encountered upon loading.
+     */
+    boolean load() throws Exception;
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]:
```java
default boolean load() throws Exception {
    return true;
}
```

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+    /**
+     * Called upon enabling the plugin. This is not called if {@link #load()} erred in some way.
+     * <p>
+     * Returns true if the plugin successfully enables, false otherwise.
+     *
+     * @return Whether the enable was successful.
+     * @throws Exception Any error encountered upon enabling.
+     */
+    boolean enable() throws Exception;
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]:
```java
default boolean enable() throws Exception {
    return true;
}
```

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+     * <ul>
+     * <li>ACF will have no more commands by this stage.</li>
+     * <li>The plugin has no more registered listeners by this stage.</li>
+     * <li>The plugin has all its tasks cancelled by this stage.</li>
+     * <li>The task chain factory will shut down after this has been called.</li>
+     * </ul>
+     *
+     * @throws Exception Any error encountered upon disabling.
+     */
+    void disable() throws Exception;
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]:
```java
default void disable() throws Exception {
}
```

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+    void shutdownSafely(boolean options);
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]: Implementation detail and should not be public to other plugins.

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    /**
+     * Gets the {@link MinevictusUtilsPlatform MV-Util} instance.
+     *
+     * @return The {@link MinevictusUtilsPlatform MV-Util} implemented instance.
+     */
+    @NotNull
+    MinevictusUtilsPlatform<?, CM> getMvUtil();
+
+    @NotNull
+    CM constructAcf();
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]: Implementation detail, see above.

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+    void shutdownSafely(boolean options);
+
+    /**
+     * Gets the {@link MinevictusUtilsPlatform MV-Util} instance.
+     *
+     * @return The {@link MinevictusUtilsPlatform MV-Util} implemented instance.
+     */
+    @NotNull
+    MinevictusUtilsPlatform<?, CM> getMvUtil();
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]: Wouldn't just having an extra generic for the platform be better?

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+    @NotNull
+    CM constructAcf();
+
+    /**
+     * Construct a database with a given database name.
+     *
+     * @return New database constructed using {@link MinevictusUtilsPlatform#prepareDatabase(String, Object)}.
+     */
+    @NotNull
+    Database constructDatabase();
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]: Implementation detail.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 15]
```diff
@@ -129,67 +137,134 @@ public final void onDisable() {
...
    * @throws Exception Any error encountered upon disabling.
    */
-  protected void disable()
+  @Override
+  public void disable()
       throws Exception {
   }
 
-  private void shutdownSafely() {
+  public void shutdownSafely() {
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Implementation detail.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 16]
```diff
@@ -129,67 +137,134 @@ public final void onDisable() {
...
+    shutdownSafely(false);
+  }
+
+  @Override
+  public void shutdownSafely(boolean options) {
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: Implementation detail.

File: bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java [hunk 17]
```diff
@@ -3,34 +3,40 @@
...
+    extends Plugin
+    implements IMvPlugin<BungeeCommandManager, Listener> {
   private final LazyValue<BungeeCommandManager> acf = new LazyValue<>(this::constructAcf);
-  private final List<PacketChannel<?>> channels = new ArrayList<>();
+  private final LazyValue<Database> database = new LazyValue<>(this::constructDatabase);
+  private final List<IPacketChannel<?, ?>> channels = new ArrayList<>();
   private Configuration configuration = null;
   private PluginErrorState errorState = null;
   private boolean enabled = false;
+  private String databaseName = getName().replace(" ", "_").toLowerCase();
```
Comment (Proximyst) [inline bungee/src/main/java/us/minevict/mvutil/bungee/MvPlugin.java]: `getName()` is not guaranteed to be populated with a value at this time. Do this in `#onLoad` using the `PluginDescriptionFile`.

File: common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+     * Registers commands to this plugin's {@link CommandManager}.
+     *
+     * @param force    Whether to overwrite other commands in the same names.
+     * @param commands The commands to register.
+     */
+    void registerCommands(boolean force, @NotNull BaseCommand... commands);
+
+    /**
+     * Registers commands to this plugin's {@link CommandManager}.
+     * <p>
```
Comment (Proximyst) [inline common/src/main/java/us/minevict/mvutil/common/IMvPlugin.java]: Remember to close this.

File: spigot/src/main/java/us/minevict/mvutil/spigot/MvPlugin.java [hunk 19]
```diff
@@ -6,31 +6,37 @@
...
+    extends JavaPlugin
+    implements IMvPlugin<PaperCommandManager, Listener> {
   private final LazyValue<PaperCommandManager> acf = new LazyValue<>(this::constructAcf);
+  private final LazyValue<Database> database = new LazyValue<>(this::constructDatabase);
   private final List<BukkitTask> tasks = new ArrayList<>();
-  private final List<PacketChannel<?>> channels = new ArrayList<>();
+  private final List<IPacketChannel<?, ?>> channels = new ArrayList<>();
   private PluginErrorState errorState = null;
   private TaskChainFactory taskChainFactory = null;
+  private String databaseName = getName().replace(" ", "_").toLowerCase();
```
Comment (Proximyst) [inline spigot/src/main/java/us/minevict/mvutil/spigot/MvPlugin.java]: `getName()` is not guaranteed to be populated yet. Do this in `#onLoad`.

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: common/src/main/java/us/minevict/mvutil/common/channel/IPacketChannel.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+ */
+public interface IPacketChannel<P, RCV> {
+  /**
+   * Send a packet through the channel.
+   *
+   * @param packet   The packet to send or null if {@link #permitsNulls()} is true.
+   * @param receiver Who should receive this packet.
+   * @return Whether a packet was successfully sent.
+   */
+  boolean sendPacket(RCV receiver, P packet);
```
Comment (NahuLD) [inline common/src/main/java/us/minevict/mvutil/common/channel/IPacketChannel.java]: I forgot to mark the receiver as NotNull here

**Comment by NahuLD** [comment]
All done, mark the requested changes as done and we merge :1234:

**Review by Proximyst — APPROVED** [review]
LGTM. Make sure to push a new tag!

**Pull request merged**